tensorpipe/test/channel/channel_test.cc

```diff
@@ -15,513 +15,568 @@
+    std::future<Error> recvFuture =
+        recvWithFuture(channel, descriptor, wrappedData.tensor());
+    Error recvError = recvFuture.get();
+    EXPECT_FALSE(recvError) << recvError.what();
+
+    // Validate contents of vector.
+    auto unwrappedData = wrappedData.unwrap();
+    for (auto i = 0; i < dataSize; i++) {
+      EXPECT_EQ(unwrappedData[i], i);
+    }
+
+    this->peers_->done(PeerGroup::kClient);
+    this->peers_->join(PeerGroup::kClient);
+
+    ctx->join();
+  }
 };
 
-TEST_P(ChannelTest, ServerToClient) {
+CHANNEL_TEST_GENERIC(ClientToServer);
+
+// TEST_P(ChannelTest, ServerToClient) {
+template <typename TTensor>
+class ServerToClientTest : public ChannelTest<TTensor> {
+  static constexpr int dataSize = 256;
+
+ public:
+  void server(std::shared_ptr<transport::Connection> conn) override {
+    std::shared_ptr<Context<TTensor>> ctx =
+        this->helper_->makeContext("server");
+    auto channel = ctx->createChannel(std::move(conn), Endpoint::kListen);
+
+    DataWrapper<TTensor> wrappedData(dataSize);
+
+    // Perform recv and wait for completion.
+    auto descriptor = this->peers_->recv(PeerGroup::kServer);
+    std::future<Error> recvFuture =
+        recvWithFuture(channel, descriptor, wrappedData.tensor());
+    Error recvError = recvFuture.get();
+    EXPECT_FALSE(recvError) << recvError.what();
+
+    // Validate contents of vector.
+    auto unwrappedData = wrappedData.unwrap();
+    for (auto i = 0; i < dataSize; i++) {
+      EXPECT_EQ(unwrappedData[i], i);
+    }
+
+    this->peers_->done(PeerGroup::kServer);
+    this->peers_->join(PeerGroup::kServer);
+
+    ctx->join();
+  }
+
+  void client(std::shared_ptr<transport::Connection> conn) override {
+    std::shared_ptr<Context<TTensor>> ctx =
+        this->helper_->makeContext("client");
+    auto channel = ctx->createChannel(std::move(conn), Endpoint::kConnect);
+
+    // Initialize with sequential values.
+    std::vector<uint8_t> data(dataSize);
+    std::iota(data.begin(), data.end(), 0);
+    DataWrapper<TTensor> wrappedData(data);
+
+    // Perform send and wait for completion.
+    std::future<std::tuple<Error, TDescriptor>> descriptorFuture;
+    std::future<Error> sendFuture;
+    std::tie(descriptorFuture, sendFuture) =
+        sendWithFuture(channel, wrappedData.tensor());
+    Error descriptorError;
+    TDescriptor descriptor;
+    std::tie(descriptorError, descriptor) = descriptorFuture.get();
+    EXPECT_FALSE(descriptorError) << descriptorError.what();
+    this->peers_->send(PeerGroup::kServer, descriptor);
+    Error sendError = sendFuture.get();
+    EXPECT_FALSE(sendError) << sendError.what();
+
+    this->peers_->done(PeerGroup::kClient);
+    this->peers_->join(PeerGroup::kClient);
+
+    ctx->join();
+  }
+};
+
+CHANNEL_TEST_GENERIC(ServerToClient);
+
+template <typename TTensor>
+class SendMultipleTensorsTest : public ChannelTest<TTensor> {
+  const int dataSize = 256 * 1024; // 256KB
```

Comment:
This was a `constexpr`, why the change?

Reply:
Ok, if it doesn't work don't bother (but maybe add a comment so that if anyone who comes around knows they can help).